new-tutorial/2_dglgraph.py

```diff
@@ -0,0 +1,211 @@
+# numerical contents). Consequently, an attribute of all the nodes or
+# edges must have the same shape. In the context of deep learning, those
+# attributes are often called *features*.
+# 
+# You can assign and retrieve node and edge features via ``ndata`` and
+# ``edata`` property.
+# 
+
+# Assign a 3-dimensional node feature vector for each node.
+g.ndata['x'] = torch.randn(6, 3)
+# Assign a 4-dimensional edge feature vector for each edge.
+g.edata['a'] = torch.randn(5, 4)
+# Assign a 5x4 node feature matrix for each node.  Node and edge features in DGL can be multi-dimensional.
+g.ndata['y'] = torch.randn(6, 5, 4)
+
+print(g.edata['a'])
+
+
+######################################################################
+# .. note::
+# 
+#    The vast development of deep learning has provided us many
+#    ways to encode various types of attributes into numerical features.
+#    Here are some general suggestions:
+# 
+#    -  For categorical attributes (e.g. gender, occupation), consider
+#       converting them to integers or one-hot encoding.
+#    -  For variable length string contents (e.g. news article, quote),
+#       consider applying a language model.
+#    -  For images, consider applying a vision model such as CNNs.
+# 
+#    You can find plenty of materials on how to encode such attributes
+#    into a tensor in the `PyTorch Deep Learning
+#    Tutorials <https://pytorch.org/tutorials/>`__.
+# 
+
+
+######################################################################
+# Graph Queries
+# -------------
+# 
+# ``DGLGraph`` object provides various methods to query a graph structure.
+# 
+
+print(g.num_nodes())
+print(g.num_edges())
+# Out degrees of the internal node
+print(g.out_degrees(0))
+# In degrees of the internal node - note that the graph is directed so the in degree should be 0.
+print(g.in_degrees(0))
+
+
+######################################################################
+# Graph Transformations
+# ---------------------
+# 
+
+
+######################################################################
+# You can also transform a graph to another graph.
+# 
+# A common transformation is to create a subgraph from the original graph:
+# 
+
+# Induce a subgraph from node 0, node 1 and node 3 from the original graph.
+sg1 = g.subgraph([0, 1, 3])
+# Induce a subgraph from edge 0, edge 1 and edge 3 from the original graph.
+sg2 = g.edge_subgraph([0, 1, 3])
+
+
+######################################################################
+# You can obtain the node/edge mapping from the subgraph to the original
+# graph by looking into the node feature ``dgl.NID`` or edge feature
+# ``dgl.EID`` in the new graph.
+# 
+
+# The original IDs of each node in sg1
+print(sg1.ndata[dgl.NID])
+# The original IDs of each edge in sg1
+print(sg1.edata[dgl.EID])
+# The original IDs of each node in sg2
+print(sg2.ndata[dgl.NID])
+# The original IDs of each edge in sg2
+print(sg2.edata[dgl.EID])
```

Comment:
Maybe feature extraction is more a common use case?

Reply:
Added below.